prompting/validators/criteria.py

```diff
@@ -0,0 +1,96 @@
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+import re
+import torch
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from typing import List
+from enum import Enum
+
+
+@dataclass
+class TaskCriterion(ABC):
+    """
+    Abstract base class for defining task-specific evaluation criteria.
+
+    Attributes:
+        text (str): Text of the criterion to be added to the prompt.
+        penalty (float): Penalty value associated with the criterion.
+    Returns:
+        torch.FloatTensor: Tensor containing the penalty values for each response.
+    """
+
+    text: str
+    penalty: float
+
+    @abstractmethod
+    def evaluate(self, completions: List[str]) -> torch.FloatTensor:
+        pass
+
+    @abstractmethod
+    def compose_text(self) -> str:
+        pass
+
+
+class TextLengthUnitEnum(Enum):
+    CHARACTERS = "characters"
+    WORDS = "words"
+    SENTENCES = "sentences"
+    PARAGRAPHS = "paragraphs"
+
+
+@dataclass
+class MatchLengthCriteria(TaskCriterion):
+    text: str = "Your response must have {target_length} {unit}."
+    penalty: float = 0.1
+    target_length: int = 100
+    unit: TextLengthUnitEnum = TextLengthUnitEnum.WORDS
+
+    def _get_completion_length(self, response: str) -> int:
+        unit_to_split_pattern = {
+            TextLengthUnitEnum.CHARACTERS: None,
+            TextLengthUnitEnum.WORDS: r" +",
+            TextLengthUnitEnum.SENTENCES: r"\. +",
+            TextLengthUnitEnum.PARAGRAPHS: r"\n\n+",
+        }
+
+        if self.unit == TextLengthUnitEnum.CHARACTERS:
+            return len(response)
+        else:
+            split_pattern = unit_to_split_pattern[self.unit]
+            return len(re.split(split_pattern, response.strip()))
+
+    def evaluate(self, completions: List[str]) -> torch.FloatTensor:
+        penalties = torch.zeros(len(completions), dtype=torch.float32)
+        for idx, completion in enumerate(completions):
+            completion_length = self._get_completion_length(completion)
+            if completion_length != self.target_length:
+                # Scales the penalty based on how close the response length is to the expected target length.
+                # If the response length is closer to the target, the penalty is reduced, ensuring that
+                # small deviations from the target length are penalized less than larger deviations.
+                penalty_scale_factor = min(
+                    abs(1 - (completion_length / self.target_length)), 1
```

Comment:
To be sure I'm following this:
```python
In [22]: np.abs(1 - (np.arange(20) / 4))
Out[22]: 
array([1.  , 0.75, 0.5 , 0.25, 0.  , 0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  , 2.25, 2.5 , 2.75, 3.  , 3.25, 3.5 , 3.75])
In [26]: np.abs(1 - (np.arange(20) / 4)).clip(0,1)
Out[26]: 
array([1.  , 0.75, 0.5 , 0.25, 0.  , 0.25, 0.5 , 0.75, 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  ])
```
When the length is correct the first function goes to zero and (increases linearly on both sides with error), and the second function increases until the error is 100% and then saturates at 1. However, when target length is, say 50 that means you don't get max penalty as long as you have length 0-100. This gives a lot of space for error. Perhaps a steeper minimum would better incentivize instruction following. Eg. squared relative error
```python
In [27]: np.abs(1 - (np.arange(20) / 4)**2).clip(0,1)
Out[27]: 
array([1.    , 0.9375, 0.75  , 0.4375, 0.    , 0.5625, 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    , 1.    ])
       ```

Reply:
![newplot (36)](https://github.com/opentensor/text-prompting/assets/6709103/0a06d6f6-5171-4659-a2b2-3f7695e48508)
This seems like a pretty good shape to me.

Reply:
![newplot (37)](https://github.com/opentensor/text-prompting/assets/6709103/0fe5a234-40a5-4edd-aec3-90693420f8fc)

Reply:
![newplot - 2023-11-01T135238 472](https://github.com/opentensor/text-prompting/assets/6709103/e05113c6-3e68-4e98-98c1-b3312c577efc)

Reply:
totally, I trust your squared relative error function, I will update the code to follow it accordingly

Reply:
Hey just to be clear I actually prefer the gaussian. Can we go with that??

Reply:
I changed accordingly to the formula you provided, it should be all good now